### PR TITLE
New file reader

### PR DESCRIPTION
**Pull request by haiqi96**
# Description
In long term, we want to be able to reuse ffi::ir_decoder for clp side ir_decoding and ir->archive compression. However, the current ffi::ir_decoder uses it's own class interface which doesn't easily fit with CLP source code. In addition, the ir->archive will require seek_back functionality on the fileReader that is not supported with our s3 layer.

This PR refactors IrBuffer to be a derived class of ReaderInterface, and also introduces a new BufferedFileReader to customized support seek back. As a side effect of the change, the compression code is also simplified,

# Validation performed
Added UnitTest for BufferedFileReader and all UnitTest passed

# Benchmark
simple benchmark for bufferedFileReader performance: compressing single 10gb hadoop log for 5 times and take the average
clp_ref: 66.956s
clp_buffered_file_reader: 64.871s

Simple benchmark to measure the improvment of ir_decoding from using string view:
BufferReader + string_copy: 86.707s
BufferReader + string_view:  77.331s

BufferedFileReader + string_copy : 88.573 s
BufferedFileReader + string_view:   81.250 s

### CODE REVIEW

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
Partial review

File: components/core/src/LibarchiveFileReader.cpp [hunk 1]
```diff
@@ -167,6 +167,60 @@ ErrorCode LibarchiveFileReader::try_read_to_delimiter (char delim, bool keep_del
...
+        if (ErrorCode_Success != error_code) {
+            data_ptr = nullptr;
+            peek_size = 0;
+            return error_code;
+        }
+    }
+
+    if (m_data_block_pos_in_file <= m_pos_in_file) {
+        // No need to to simulate reading '\0' before the start of the data block
+        // simply return a const pointer to the current data block
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveFileReader.cpp:191]:
```suggestion
        // Position in the file is within the data block, so we can directly
        // return a pointer to the data block
```

File: components/core/src/LibarchiveFileReader.cpp [hunk 2]
```diff
@@ -167,6 +167,60 @@ ErrorCode LibarchiveFileReader::try_read_to_delimiter (char delim, bool keep_del
...
+        peek_size = std::min(size_to_peek, m_data_block_length - m_pos_in_data_block);
+        data_ptr = reinterpret_cast<const char*>(m_data_block);
+        return ErrorCode_Success;
+    }
+
+    // There are sparse bytes before the data block, so the pos in data block
+    // must be 0
+    assert(m_pos_in_data_block != 0);
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveFileReader.cpp:199]: Comment is the opposite of the assertion?

File: components/core/src/LibarchiveFileReader.cpp [hunk 3]
```diff
@@ -167,6 +167,60 @@ ErrorCode LibarchiveFileReader::try_read_to_delimiter (char delim, bool keep_del
...
+    auto num_sparse_bytes = m_data_block_pos_in_file - m_pos_in_file;
+    peek_size = std::min(num_sparse_bytes + m_data_block_length, size_to_peek);
+
+    // resize the local buffer if necessary
+    if (m_data_for_peek.size() < peek_size) {
+        m_data_for_peek.resize(peek_size);
+    }
+    data_ptr = reinterpret_cast<const char*>(m_data_for_peek.data());
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveFileReader.cpp:207]: You don't need a cast.

File: components/core/src/LibarchiveFileReader.cpp [hunk 4]
```diff
@@ -167,6 +167,60 @@ ErrorCode LibarchiveFileReader::try_read_to_delimiter (char delim, bool keep_del
     return ErrorCode_Success;
 }
 
+ErrorCode LibarchiveFileReader::peek_data_block (size_t size_to_peek, const char*& data_ptr,
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveFileReader.cpp:170]: `size_to_peek` seems unnecessary since for sparse files, we could just return a statically sized buffer with nulls. Allowing the caller to dictate how much we peek seems more like a read method which creates unnecessary complexity imo.

File: components/core/src/LibarchiveFileReader.cpp [hunk 5]
```diff
@@ -167,6 +167,60 @@ ErrorCode LibarchiveFileReader::try_read_to_delimiter (char delim, bool keep_del
...
+            peek_size = 0;
+            return error_code;
+        }
+    }
+
+    if (m_data_block_pos_in_file <= m_pos_in_file) {
+        // No need to to simulate reading '\0' before the start of the data block
+        // simply return a const pointer to the current data block
+        peek_size = std::min(size_to_peek, m_data_block_length - m_pos_in_data_block);
+        data_ptr = reinterpret_cast<const char*>(m_data_block);
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveFileReader.cpp:193]: static_cast

File: components/core/src/LibarchiveFileReader.cpp [hunk 6]
```diff
@@ -167,6 +167,60 @@ ErrorCode LibarchiveFileReader::try_read_to_delimiter (char delim, bool keep_del
...
+    peek_size = std::min(num_sparse_bytes + m_data_block_length, size_to_peek);
+
+    // resize the local buffer if necessary
+    if (m_data_for_peek.size() < peek_size) {
+        m_data_for_peek.resize(peek_size);
+    }
+    data_ptr = reinterpret_cast<const char*>(m_data_for_peek.data());
+
+    if (size_to_peek < num_sparse_bytes) {
+        memset(m_data_for_peek.data(), '\0', size_to_peek);
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveFileReader.cpp:210]:
```suggestion
        std::fill(m_data_for_peek.begin(), m_data_for_peek.end(), '\0');
```

File: components/core/src/LibarchiveFileReader.hpp [hunk 7]
```diff
@@ -69,6 +71,18 @@ class LibarchiveFileReader : public ReaderInterface {
...
+     * if no enough data is available in the next data block, a smaller peek
+     * size will be returned
+     * @param size_to_peek
+     * @param data_ptr
+     * @param peek_size Return the number of bytes peeked by reference
+     * @return ErrorCode_EndOfFile on EOF
+     * @return ErrorCode_Failure on failure
+     * @return ErrorCode_Success on success
+     */
+    [[nodiscard]] ErrorCode peek_data_block(size_t size_to_peek, const char*& data_ptr, size_t& peek_size);
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveFileReader.hpp:85]: Need a newline after this

File: components/core/src/LibarchiveReader.hpp [hunk 8]
```diff
@@ -43,15 +43,13 @@ class LibarchiveReader {
 
     // Methods
     /**
-     * Tries to open the archive or compressed file contained across the given buffer and FileReader
-     * @param buffer_length
-     * @param buffer
+     * Tries to open the archive or compressed file contained in the FileReader
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveReader.hpp:46]:
```suggestion
     * Tries to open the archive or compressed file contained in the reader
```

File: components/core/src/LibarchiveReader.hpp [hunk 9]
```diff
@@ -43,15 +43,13 @@ class LibarchiveReader {
...
      * @param file_reader
      * @param path_if_compressed_file Path to use if the data is a single compressed file
      * @return ErrorCode_Success on success
      * @return ErrorCode_Failure on failure
      */
-    ErrorCode try_open (size_t buffer_length, const char* buffer, FileReader& file_reader, const std::string& path_if_compressed_file);
+    ErrorCode try_open (ReaderInterface& file_reader, const std::string& path_if_compressed_file);
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveReader.hpp:52]:
```suggestion
    ErrorCode try_open (ReaderInterface& reader, const std::string& path_if_compressed_file);
```

File: components/core/src/LibarchiveReader.hpp [hunk 10]
```diff
@@ -160,7 +158,7 @@ class LibarchiveReader {
     struct archive_entry* m_archive_entry;
 
     std::vector<char> m_buffer;
-    FileReader* m_file_reader;
+    ReaderInterface* m_file_reader;
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveReader.hpp:161]:
```suggestion
    ReaderInterface* m_reader;
```

File: components/core/src/LibarchiveReader.hpp [hunk 11]
```diff
@@ -160,7 +158,7 @@ class LibarchiveReader {
...
     bool m_initial_buffer_content_exhausted;
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveReader.hpp:162]: No longer used.

File: components/core/src/LibarchiveReader.cpp [hunk 12]
```diff
@@ -186,53 +180,31 @@ ErrorCode LibarchiveReader::libarchive_read_callback (const void** buffer, size_
...
-    auto error_code = m_file_reader->try_get_pos(pos);
-    if (ErrorCode_Success != error_code) {
-        return error_code;
-    }
-
-    // Calculate desired position, ensuring its within the file
-    size_t desired_pos = pos + num_bytes_to_skip;
-    struct stat stat_buffer = {};
-    error_code = m_file_reader->try_fstat(stat_buffer);
+    constexpr size_t cTargetBufferLength = 4096;
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveReader.cpp:183]: Let's retain the newline before this.

File: components/core/src/LibarchiveReader.cpp [hunk 13]
```diff
@@ -186,53 +180,31 @@ ErrorCode LibarchiveReader::libarchive_read_callback (const void** buffer, size_
...
+}
 
-    // Seek to desired position
-    error_code = m_file_reader->try_seek_from_begin(desired_pos);
-    if (ErrorCode_Success != error_code) {
+ErrorCode LibarchiveReader::libarchive_skip_callback (off_t num_bytes_to_skip,
+                                                      size_t& num_bytes_skipped) {
+
+    // skip bytes by simply reading data into a temporary buffer
+    std::vector<char> temporary_read_buffer;
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveReader.cpp:200]: Reading into an empty vector?

File: components/core/src/LibarchiveReader.cpp [hunk 14]
```diff
@@ -186,53 +180,31 @@ ErrorCode LibarchiveReader::libarchive_read_callback (const void** buffer, size_
...
+        m_buffer.resize(num_bytes_read);
     }
+    *buffer = m_buffer.data();
+    return ErrorCode_Success;
+}
 
-    // Seek to desired position
-    error_code = m_file_reader->try_seek_from_begin(desired_pos);
-    if (ErrorCode_Success != error_code) {
+ErrorCode LibarchiveReader::libarchive_skip_callback (off_t num_bytes_to_skip,
```
Comment (kirkrodrigues) [inline components/core/src/LibarchiveReader.cpp:196]: We don't need this callback anymore since we're not doing anything more efficient than reading (which is what libarchive will do when there's no skip callback).

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 15]
```diff
@@ -95,7 +97,7 @@ bool encode_message (epoch_time_ms_t timestamp, string_view message, string& log
  * encoded_variable_t == four_byte_encoded_variable_t
  */
 template <typename encoded_variable_t>
-IRErrorCode decode_next_message (IrBuffer& ir_buf, string& message, epoch_time_ms_t& decoded_ts);
+IRErrorCode decode_next_message (BufferReader& ir_buf, string& message, epoch_time_ms_t& decoded_ts);
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:100]:
```suggestion
IRErrorCode decode_next_message (BufferReader& reader, string& message, epoch_time_ms_t& decoded_ts);
```

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 16]
```diff
@@ -204,40 +206,46 @@ static void set_timestamp_info (const nlohmann::json& metadata_json, TimestampIn
 
 TEST_CASE("get_encoding_type", "[ffi][get_encoding_type]") {
     bool is_four_bytes_encoding;
-
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:207]: Why delete the newline?

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 17]
```diff
@@ -204,40 +206,46 @@ static void set_timestamp_info (const nlohmann::json& metadata_json, TimestampIn
...
     // Test eight-byte encoding
     vector<int8_t> eight_byte_encoding_vec{EightByteEncodingMagicNumber,
                                            EightByteEncodingMagicNumber + MagicNumberLength};
 
-    IrBuffer eight_byte_ir_buffer(eight_byte_encoding_vec.data(),
-                                  eight_byte_encoding_vec.size());
-    REQUIRE(get_encoding_type(eight_byte_ir_buffer, is_four_bytes_encoding) ==
+    // Test eight-byte encoding
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:213]: Why repeat the comment?

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 18]
```diff
@@ -204,40 +206,46 @@ static void set_timestamp_info (const nlohmann::json& metadata_json, TimestampIn
...
+    REQUIRE(get_encoding_type(four_byte_encoding_buffer, is_four_bytes_encoding) ==
             IRErrorCode::IRErrorCode_Success);
     REQUIRE(match_encoding_type<four_byte_encoded_variable_t>(is_four_bytes_encoding));
 
-    // Test error on empty and incomplete ir_buffer
-    const vector<int8_t> empty_ir_vec;
-    IrBuffer empty_ir_buffer(empty_ir_vec.data(), empty_ir_vec.size());
-    REQUIRE(get_encoding_type(empty_ir_buffer, is_four_bytes_encoding) ==
-            IRErrorCode::IRErrorCode_Incomplete_IR);
+    // Test error on incomplete ir_buffer
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:234]: We don't need to remove the empty buffer test anymore right?

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 19]
```diff
@@ -256,32 +264,46 @@ TEMPLATE_TEST_CASE("decode_preamble", "[ffi][decode_preamble]", four_byte_encode
...
-    int8_t metadata_type{0};
+    encoded_tag_t metadata_type{0};
     size_t metadata_pos{0};
     uint16_t metadata_size{0};
-    REQUIRE(decode_preamble(preamble_buffer, metadata_type, metadata_pos, metadata_size) ==
+    REQUIRE(decode_preamble(encoding_buffer, metadata_type, metadata_pos, metadata_size) ==
+            IRErrorCode::IRErrorCode_Success);
+    REQUIRE(encoded_preamble_end_pos == encoding_buffer.get_pos());
+    
+    auto json_metadata_ptr = reinterpret_cast<char*>(ir_buf.data() + metadata_pos);
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:285]: Why change the `metadata_ptr` line?

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 20]
```diff
@@ -256,32 +264,46 @@ TEMPLATE_TEST_CASE("decode_preamble", "[ffi][decode_preamble]", four_byte_encode
...
+    string_view json_metadata_ref {json_metadata_ptr, metadata_size};
+
+    // Test if preamble can be decoded by the string copy method
+    std::vector<int8_t> json_metadata_vec;
+    encoding_buffer.seek_from_begin(MagicNumberLength);
+    REQUIRE(decode_preamble(encoding_buffer, metadata_type, json_metadata_vec) ==
             IRErrorCode::IRErrorCode_Success);
-    REQUIRE(encoded_preamble_end_pos == preamble_buffer.get_cursor_pos());
+    string_view json_metadata_copied { reinterpret_cast<const char*>(json_metadata_vec.data()),
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:293]: `size_checked_pointer_cast`

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 21]
```diff
@@ -256,32 +264,46 @@ TEMPLATE_TEST_CASE("decode_preamble", "[ffi][decode_preamble]", four_byte_encode
...
     uint16_t metadata_size{0};
-    REQUIRE(decode_preamble(preamble_buffer, metadata_type, metadata_pos, metadata_size) ==
+    REQUIRE(decode_preamble(encoding_buffer, metadata_type, metadata_pos, metadata_size) ==
+            IRErrorCode::IRErrorCode_Success);
+    REQUIRE(encoded_preamble_end_pos == encoding_buffer.get_pos());
+    
+    auto json_metadata_ptr = reinterpret_cast<char*>(ir_buf.data() + metadata_pos);
+    string_view json_metadata_ref {json_metadata_ptr, metadata_size};
+
+    // Test if preamble can be decoded by the string copy method
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:288]: I think this would be clearer if you finished testing the first `decode_preamble` method and then tested the second one.

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 22]
```diff
@@ -319,24 +343,29 @@ TEMPLATE_TEST_CASE("decode_next_message_general", "[ffi][decode_next_message]",
     const size_t encoded_message_end_pos = ir_buf.size();
     const size_t encoded_message_start_pos = 0;
 
-    IrBuffer encoded_message_buffer(ir_buf.data(), ir_buf.size());
+    // Test if message can be decoded properly
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:346]: Comment in the wrong place?

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 23]
```diff
@@ -204,40 +206,46 @@ static void set_timestamp_info (const nlohmann::json& metadata_json, TimestampIn
...
-
     // Test eight-byte encoding
     vector<int8_t> eight_byte_encoding_vec{EightByteEncodingMagicNumber,
                                            EightByteEncodingMagicNumber + MagicNumberLength};
 
-    IrBuffer eight_byte_ir_buffer(eight_byte_encoding_vec.data(),
-                                  eight_byte_encoding_vec.size());
-    REQUIRE(get_encoding_type(eight_byte_ir_buffer, is_four_bytes_encoding) ==
+    // Test eight-byte encoding
+    BufferReader eight_byte_encoding_buffer{
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:214]: These name changes seem unnecessary, especially given that you're changing names in the opposite direction (encoding -> ir) in other places.

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 24]
```diff
@@ -364,22 +393,29 @@ TEST_CASE("message_decode_error", "[ffi][decode_next_message]")
     epoch_time_ms_t timestamp;
 
     // Test if a trailing escape triggers a decoder error
-    auto ir_with_extra_escape{ir_buf};
+    auto ir_with_extra_escape {ir_buf};
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:396]: Unnecessary change

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 25]
```diff
@@ -444,15 +481,15 @@ TEMPLATE_TEST_CASE("decode_ir_complete", "[ffi][decode_next_message]",
...
     uint16_t metadata_size;
     REQUIRE(decode_preamble(complete_encoding_buffer, metadata_type, metadata_pos, metadata_size) ==
             IRErrorCode::IRErrorCode_Success);
-    REQUIRE(encoded_preamble_end_pos == complete_encoding_buffer.get_cursor_pos());
+    REQUIRE(encoded_preamble_end_pos == complete_encoding_buffer.get_pos());
 
-    string_view json_metadata;
-    REQUIRE(complete_encoding_buffer.try_read(json_metadata, metadata_size));
+    auto json_metadata_ptr = reinterpret_cast<char*>(ir_buf.data() + metadata_pos);
+    string_view json_metadata {json_metadata_ptr, metadata_size};
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:492]:
```suggestion
    char* json_metadata_ptr{size_checked_pointer_cast<char>(ir_buf.data() + metadata_pos)};
    string_view json_metadata{json_metadata_ptr, metadata_size};
```

File: components/core/CMakeLists.txt [hunk 26]
```diff
@@ -624,6 +632,10 @@ target_compile_features(clo
         )
 
 set(SOURCE_FILES_unitTest
+        src/BufferReader.cpp
```
Comment (kirkrodrigues) [inline components/core/CMakeLists.txt:635]: Wrong order

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
* Rebase your changes to main so you can run clang-format on more files you've changed
* Please run clang-format on your new files
* Please try to resolve as many clang-tidy warnings as you can

File: components/core/src/BufferedFileReader.cpp [hunk 27]
```diff
@@ -0,0 +1,385 @@
...
+}
+
+size_t BufferedFileReader::quantize_to_buffer_size (size_t size) const {
+    if (size == 0) {
+        return 0;
+    }
+    return (1 + ((size - 1) / m_base_buffer_size)) * m_base_buffer_size;
+}
+
+ErrorCode BufferedFileReader::refill_reader_buffer (size_t num_bytes_to_refill) {
```
Comment (kirkrodrigues) [inline components/core/src/BufferedFileReader.cpp:284]: I feel like this method would be more clear with the following structure:
```
if checkpoint:
    grow the buffer if necessary
    calculate the number of bytes to read
    calculate the destination for the read data
        buf_begin_pos = m_buf_reader->get_pos()
else:
    if we haven't filled the current base-buffer-size:
        calculate the number of bytes to read
        calculate the destination for the read data
        buf_begin_pos = m_buf_reader->get_pos()
    else:
        advance the buffer's position
        set the number of bytes to read to base buffer size
        set the destination for the read data to 0
        buf_begin_pos = 0

read num_bytes_to_read from the file
if there's an error and it's not EOF:
    return the error

Initialize the buffer reader with the variables you calculated above

return the error_code
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/tests/test-ir_encoding_methods.cpp (hunk 25 above)
Comment (haiqi96) [reply components/core/tests/test-ir_encoding_methods.cpp:492]: why not use auto* ?

File: components/core/src/BufferedFileReader.cpp [hunk 28]
```diff
@@ -0,0 +1,385 @@
...
+            m_buffer_reader->try_seek_from_begin(get_corresponding_offset(pos))) {
+            m_file_pos = pos;
+            highest_read_pos = std::max(highest_read_pos, m_file_pos);
+            return ErrorCode_Success;
+        }
+
+        size_t num_bytes_to_refill = pos - get_buffer_end_pos();
+        auto error_code = refill_reader_buffer(num_bytes_to_refill);
+        if (ErrorCode_EndOfFile == error_code) {
+            throw OperationFailed(ErrorCode_EndOfFile, __FILENAME__, __LINE__);
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/BufferedFileReader.cpp:91]: I think it's a valid use case to try and seek to a file_pos that's beyond the size of the file, even though we may not have it right now. For instance, we may be skipping around in a truncated IR file, but we won't know it's truncated when we start reading/seeking in it. The caller shouldn't have to handle the truncation as an exception along with with error codes for other errors.

File: components/core/src/BufferedFileReader.cpp [hunk 29]
```diff
@@ -0,0 +1,385 @@
...
+
+    if (m_checkpoint_pos.has_value() == false) {
+        if (pos < m_file_pos) {
+            return ErrorCode_Failure;
+        }
+        if (ErrorCode_Success ==
+            m_buffer_reader->try_seek_from_begin(get_corresponding_offset(pos))){
+            m_file_pos = pos;
+            highest_read_pos = std::max(highest_read_pos, m_file_pos);
+            return ErrorCode_Success;
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/BufferedFileReader.cpp:66]: Yeah

File: components/core/src/BufferReader.cpp [hunk 30]
```diff
@@ -0,0 +1,81 @@
...
+        ret_code = ErrorCode_Success;
+    } else {
+        delim_pos = m_internal_buf_size;
+        ret_code = ErrorCode_EndOfFile;
+    }
+    // append to strings
+    length = delim_pos - m_internal_buf_pos;
+    if (false == keep_delimiter && delim == m_internal_buf[delim_pos - 1]) {
+        --length;
+    }
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/BufferReader.cpp:77]: I'm saying you could restructure the method so that you only increment the length if delim_ptr != nullptr before you get to this line.

File: components/core/tests/test-ir_encoding_methods.cpp (hunk 25 above)
Comment (kirkrodrigues) [reply components/core/tests/test-ir_encoding_methods.cpp:492]: Yeah, you should.

File: components/core/src/BufferedFileReader.cpp [hunk 31]
```diff
@@ -0,0 +1,385 @@
...
+        auto offset = lseek(m_fd, pos, SEEK_SET);
+        if (offset == -1) {
+            return ErrorCode_errno;
+        }
+        m_buffer_reader.emplace(m_buffer.get(), 0);
+        m_buffer_begin_pos = pos;
+    } else {
+        if (pos < m_checkpoint_pos) {
+            return ErrorCode_Failure;
+        } else if (pos < m_file_pos) {
```
[the remark this answers is not in the document]
Comment (haiqi96) [reply components/core/src/BufferedFileReader.cpp:78]: I don't think I follow this comment. I do agree that else if is redundant though

File: components/core/src/BufferedFileReader.cpp [hunk 32]
```diff
@@ -0,0 +1,385 @@
...
+        size_t num_bytes_to_refill = pos - get_buffer_end_pos();
+        auto error_code = refill_reader_buffer(num_bytes_to_refill);
+        if (ErrorCode_EndOfFile == error_code) {
+            throw OperationFailed(ErrorCode_EndOfFile, __FILENAME__, __LINE__);
+        }
+        if (ErrorCode_Success != error_code) {
+            return error_code;
+        }
+        if (ErrorCode_Success != m_buffer_reader->try_seek_from_begin(get_corresponding_offset(pos))) {
+            throw OperationFailed(ErrorCode_EndOfFile, __FILENAME__, __LINE__);
```
[the remark this answers is not in the document]
Comment (haiqi96) [reply components/core/src/BufferedFileReader.cpp:113]: This should not be an excpetion anymore. it should be Errorcode Truncated instead

File: components/core/src/BufferedFileReader.cpp [hunk 33]
```diff
@@ -0,0 +1,385 @@
...
+    if (m_checkpoint_pos.has_value() == false) {
+        if (pos < m_file_pos) {
+            return ErrorCode_Failure;
+        }
+        if (ErrorCode_Success ==
+            m_buffer_reader->try_seek_from_begin(get_corresponding_offset(pos))){
+            m_file_pos = pos;
+            highest_read_pos = std::max(highest_read_pos, m_file_pos);
+            return ErrorCode_Success;
+        }
```
[the remark this answers is not in the document]
Comment (haiqi96) [reply components/core/src/BufferedFileReader.cpp:85]: Refactored this function

File: components/core/src/BufferedFileReader.cpp (hunk 27 above)
Comment (haiqi96) [reply components/core/src/BufferedFileReader.cpp:291]: refactored in a way similar to the structure you proposed but I am not sure if it is what you expect

**Comment by kirkrodrigues** [comment]
A future PR should replace the pointer arithmetic with `gsl::span`

**Comment by haiqi96** [comment]
TODO: use gsl::span to replace pointer arithmetic

File: components/core/src/BufferedFileReader.cpp [hunk 34]
```diff
@@ -0,0 +1,385 @@
...
+            throw OperationFailed(error_code, __FILENAME__, __LINE__);
+        }
+    }
+}
+
+void BufferedFileReader::close () {
+    if (-1 != m_fd) {
+        // NOTE: We don't check errors for fclose since it seems
+        // the only reason it could fail is if it was interrupted by a signal
+        ::close(m_fd);
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/BufferedFileReader.cpp:231]: close isn't an overridden method is it?

File: components/core/src/BufferReader.cpp [hunk 35]
```diff
@@ -0,0 +1,100 @@
...
+    char const* buffer_head = m_internal_buf + m_internal_buf_pos;
+    char const* delim_ptr
+            = static_cast<char const*>(memchr(buffer_head, delim, remaining_data_size));
+
+    size_t delim_pos{0};
+    if (delim_ptr != nullptr) {
+        delim_pos = (delim_ptr - m_internal_buf) + 1;
+        num_bytes_read = delim_pos - m_internal_buf_pos;
+        if (false == keep_delimiter && delim == m_internal_buf[delim_pos - 1]) {
+            --num_bytes_read;
```
Comment (kirkrodrigues) [inline components/core/src/BufferReader.cpp:89]: The number of bytes read should be the same regardless of whether we keep the delimiter in the returned string or not.

File: components/core/src/BufferedFileReader.cpp (hunk 34 above)
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/BufferedFileReader.cpp]: Sorry, this was my mistake. We don't need to check for errors from close when reading files.

**Review by davidlion — APPROVED** [review]

**Review by kirkrodrigues — APPROVED** [review]
